src/main/scala/zio/interop/reactiveStreams/Adapters.scala

```diff
@@ -46,10 +40,153 @@ object Adapters {
     } yield (error, demandUnfoldSink(subscriber, demand))
 
   def publisherToStream[A](publisher: Publisher[A], bufferSize: Int): ZStream[Any, Throwable, A] =
-    Stream.unwrap(
-      QueueSubscriber.make[A](bufferSize).flatMap {
-        case (subscriber, stream) => UIO(publisher.subscribe(subscriber)).const(stream)
-      }
+    new ZStream(
+      for {
+        (q, subscription, completion, subscriber) <- makeSubscriber[A](bufferSize).toManaged_
+        _                                         <- UIO(publisher.subscribe(subscriber)).toManaged_
+        sub <- subscription.await.interruptible
+                .toManaged(sub => UIO { sub.cancel() }.whenM(completion.isDone.map(!_)))
+        process <- process(q, sub, completion)
+      } yield process
     )
 
+  def sinkToSubscriber[R, R1 <: R, A1, A, B](
+    sink: ZSink[R, Throwable, A1, A, B],
+    bufferSize: Int
+  ): ZManaged[R1, Throwable, (Subscriber[A], IO[Throwable, B])] =
+    for {
+      (q, subscription, completion, subscriber) <- makeSubscriber[A](bufferSize).toManaged_
+      result                                    <- Promise.make[Throwable, B].toManaged_
+      _ <- new ZStream(for {
+            sub <- subscription.await.interruptible
+                    .toManaged(sub => UIO { sub.cancel() }.whenM(completion.isDone.map(!_)))
+            process <- process(q, sub, completion)
+          } yield process)
+            .run(sink)
+            .to(result)
+            .interruptible // `to` is not interruptible by default
```

Comment:
It actually is, but it's running in a fiber that is forked from a fiber running as a ZManaged acquisition step, which means it inherited uninterruptibility. :-)